### PR TITLE
Initial gaps support

### PR DESCRIPTION
**Pull request by stapelberg**
I still have a few refactorings I want to give a try before we should merge this, but I think the current state is good enough for a first pull request review :) 

related to https://github.com/i3/i3/issues/3724

### CODE REVIEW

File: docs/userguide [hunk 1]
```diff
@@ -1426,6 +1426,95 @@ tiling_drag modifier titlebar
...
+gaps outer|horizontal|vertical|top|left|bottom|right <gap_size_in_px>
+
+# Inner and outer gaps for all windows on a specific workspace.
+# <ws> can be a workspace number or name.
+workspace <ws> gaps inner <gap_size_in_px>
+workspace <ws> gaps outer|horizontal|vertical|top|left|bottom|right <gap_size_in_px>
+
+# Enabling “Smart Gaps” means no gaps will be shown when there is
+# precisely one window or split container on the workspace.
+smart_gaps on|off|inverse_outer
```
Comment (Airblader) [inline docs/userguide:1473]: I think we need to explain the difference between on and inverse_outer here.

File: parser-specs/config.spec [hunk 2]
```diff
@@ -65,6 +68,32 @@ state IGNORE_LINE:
...
+
+# smart_borders true|false
+# smart_borders no_gaps
+state SMART_BORDERS:
+  enabled = '1', 'yes', 'true', 'on', 'enable', 'active'
+      -> call cfg_smart_borders($enabled)
+  enabled = 'no_gaps'
+      -> call cfg_smart_borders($enabled)
+
+# smart_gaps on|off
```
Comment (Airblader) [inline parser-specs/config.spec]: This comment needs to be updated, too.

File: docs/userguide (hunk 1 above)
Comment (stapelberg) [reply docs/userguide:1473]: Done

File: parser-specs/config.spec (hunk 2 above)
Comment (stapelberg) [reply parser-specs/config.spec]: Done

**Comment by stapelberg** [comment]
Pushed the other commits I had in mind: moving code to gaps.c where applicable, allowing optional “px” suffix, getting rid of `#define` in cmd_gaps and test cases for the `gaps` command.

Please take another look. If this looks reasonable, we could merge this state and do further cleanups/improvements in separate PRs :)

File: src/commands.c [hunk 3]
```diff
@@ -2375,114 +2375,171 @@ void cmd_debuglog(I3_CMD, const char *argument) {
...
+        reset = true;
+    } else {
+        ELOG("Invalid mode %s when changing gaps", mode);
+        return false;
+    }
+
+    /* See https://github.com/Airblader/i3/issues/262 */
+    int min_value = 0;
+    const bool is_outer = get(&(config.gaps)) != gaps_inner(&(config.gaps));
+    // if (strcmp(type, "inner") != 0) {
```
Comment (Airblader) [inline src/commands.c]: Leftover commented out code?

**Review by Airblader — APPROVED** [review]
LGTM % one nit. I have to admit that I find the define-based implementation easier to follow, but I might be biased there, am not really as experienced in C as my involvement in i3 would suggest, and it's not a big difference either way so happy to roll with this.

Thanks so much for taking care of this. We'll make a lot of people happy.

File: src/commands.c (hunk 3 above)
Comment (stapelberg) [reply src/commands.c]: Done

**Pull request merged**